components/datetime/src/neo_marker.rs

```diff
@@ -8,199 +8,358 @@ use core::marker::PhantomData;
+    }
+}
+
+impl<C: Calendar, A: AsCalendar<Calendar = C>> IsAnyCalendarKind for DateTime<A> {
+    fn is_any_calendar_kind(&self, any_calendar_kind: AnyCalendarKind) -> bool {
+        self.date.calendar().any_calendar_kind() == Some(any_calendar_kind)
+    }
+}
+
+impl IsAnyCalendarKind for Time {
+    fn is_any_calendar_kind(&self, _: AnyCalendarKind) -> bool {
+        true
+    }
+}
+
+/// An input associated with a specific calendar.
+#[derive(Debug, Copy, Clone)]
+pub struct NeoTypedInput<C> {
     /// The phantom calendar marker.
     _calendar: PhantomData<C>,
 }
 
-impl<C> NeoDateInputFields<C> {
-    /// Constructor for [`NeoDateInputFields`] with all required fields.
-    ///
-    /// Optional fields such as [`Self::any_calendar_kind`] should be populated manually.
-    pub fn new_with_year_month_day_weekday(
-        year: FormattableYear,
-        month: FormattableMonth,
-        day_of_month: DayOfMonth,
-        day_of_week: IsoWeekday,
-    ) -> Self {
-        Self {
-            year,
-            month,
-            day_of_month,
-            day_of_week,
-            any_calendar_kind: None,
+impl<C: Calendar, A: AsCalendar<Calendar = C>> NeoGetField<NeoTypedInput<C>> for Date<A> {
+    #[inline]
+    fn get_field(&self) -> NeoTypedInput<C> {
+        NeoTypedInput {
             _calendar: PhantomData,
         }
     }
 }
 
-impl<C: Calendar, A: AsCalendar<Calendar = C>> From<&Date<A>> for NeoDateInputFields<C> {
-    fn from(value: &Date<A>) -> Self {
-        Self {
-            year: value.year(),
-            month: value.month(),
-            day_of_month: value.day_of_month(),
-            day_of_week: value.day_of_week(),
-            any_calendar_kind: value.calendar().any_calendar_kind(),
+impl<C> NeoGetField<NeoTypedInput<C>> for Time {
+    #[inline]
+    fn get_field(&self) -> NeoTypedInput<C> {
+        NeoTypedInput {
             _calendar: PhantomData,
         }
     }
 }
 
-impl<C: Calendar, A: AsCalendar<Calendar = C>> From<&DateTime<A>> for NeoDateInputFields<C> {
-    fn from(value: &DateTime<A>) -> Self {
-        Self {
-            year: value.date.year(),
-            month: value.date.month(),
-            day_of_month: value.date.day_of_month(),
-            day_of_week: value.date.day_of_week(),
-            any_calendar_kind: value.date.calendar().any_calendar_kind(),
+impl<C: Calendar, A: AsCalendar<Calendar = C>> NeoGetField<NeoTypedInput<C>> for DateTime<A> {
+    #[inline]
+    fn get_field(&self) -> NeoTypedInput<C> {
+        NeoTypedInput {
             _calendar: PhantomData,
         }
     }
 }
 
-/// Input data fields required for formatting weeks of the year.
-#[derive(Debug, Copy, Clone)]
-#[non_exhaustive]
-pub struct NeoWeekInputFields {
-    /// Information on the position of the day within the year.
-    pub day_of_year_info: DayOfYearInfo,
+/// A type that can return a certain field `T`.
+pub trait NeoGetField<T> {
```

Comment:
praise: I like this trait name.


thought: We may however want to call it GetDateTimeField or something once we remove the Neo. No strong feeling.